### PR TITLE
DT-19419 Add filter for search method

### PR DESCRIPTION
**Pull request by tayloranderson-wf**
Allow more granular filtering than simply a container rectangle

### CODE REVIEW

File: lib/src/r_tree/r_tree.dart [hunk 1]
```diff
@@ -51,12 +51,17 @@ class RTree<E> {
 
   // Returns all items whose rectangles overlap the @searchRect
   //  Note: Rectangles that share only a border are not considered to overlap
-  Iterable<RTreeDatum<E>> search(Rectangle searchRect) {
+  Iterable<RTreeDatum<E>> search(Rectangle searchRect,
+      [bool Function(E item) shouldInclude]) {
```
Comment (jacobfeddersen-wf) [inline lib/src/r_tree/r_tree.dart:55]: Kind of personal preference, but I think named optional parameters are easier to maintain than optional positional parameters

**Comment by aviary-wf** [comment]
## Security Insights
No security relevant content was detected by automated scans.

## Action Items
* Review PR for [security impact](https://wiki.atl.workiva.net/display/SECURITY/Development+Security+Review+Guidelines); comment "security review required" if needed or unsure
* Verify [`aviary.yaml`](https://dev.workiva.net/docs/teams/information-security/aviary#configuring-aviary) coverage of security relevant code
---
_Questions or Comments? Reach out on Slack: #support-infosec._

File: lib/src/r_tree/r_tree.dart (hunk 1 above)
Comment (tayloranderson-wf) [reply lib/src/r_tree/r_tree.dart]: I agree 😄  Much. More. Better.

File: lib/src/r_tree/r_tree.dart [hunk 2]
```diff
@@ -51,12 +51,17 @@ class RTree<E> {
 
   // Returns all items whose rectangles overlap the @searchRect
   //  Note: Rectangles that share only a border are not considered to overlap
-  Iterable<RTreeDatum<E>> search(Rectangle searchRect) {
+  Iterable<RTreeDatum<E>> search(Rectangle searchRect,
+      {bool Function(E item) shouldInclude}) {
+    if (shouldInclude == null) {
+      shouldInclude = (_) => true;
+    }
```
Comment (jacobfeddersen-wf) [inline lib/src/r_tree/r_tree.dart:58]: Also this could be `shouldInclude ??= (_) => true`

**Review by jacobfeddersen-wf — APPROVED** [review]

**Review by derekolsthoorn-wk — DISMISSED** [review]

File: lib/src/r_tree/leaf_node.dart [hunk 3]
```diff
@@ -28,8 +28,10 @@ class LeafNode<E> extends Node<E> {
     return new LeafNode<E>(branchFactor);
   }
 
-  Iterable<RTreeDatum<E>> search(Rectangle searchRect) {
-    return _items.where((RTreeDatum<E> item) => item.overlaps(searchRect));
+  Iterable<RTreeDatum<E>> search(
+      Rectangle searchRect, bool Function(E item) shouldInclude) {
+    return _items.where((RTreeDatum<E> item) =>
+        shouldInclude(item.value) && item.overlaps(searchRect));
```
Comment (jacobfeddersen-wf) [inline lib/src/r_tree/leaf_node.dart:34]: Do you think it matters for performance which of these checks is first, for short circuiting out if it is false? Do you think we more likely to hit lots of items that don't overlap with the search rect, or lots of items that don't meet our filter criteria?

Comment (tayloranderson-wf) [reply lib/src/r_tree/leaf_node.dart]: That's a great question. We'll probably hit more items that do not overlap the searchRect, however, that overlaps method is just a little bit of simply integer comparisons, so it'll be fast. The filter argument could be as bad as the implementation provided by the caller 😄  

I'll switch the order

**Review by joshbeam-wf — APPROVED** [review]

**Comment by jacobfeddersen-wf** [comment]
QA+1
We weren't able to test with the big document with tons of links because wk-dev is down, but smoke tested applying overlapping formats and verified that various format changes in various namespaces work properly still on the client after consuming this change

**Pull request merged**